backend/src/validators/applicationValidator.ts

```diff
@@ -1,4 +1,86 @@
-import { body } from "express-validator";
+import { body, param, query } from "express-validator";
+import Company from "src/models/Company";
 
-export const createApplicationValidator = {};
-export const updateApplicationValidator = {};
+const validateId = param("id")
+  .isMongoId()
+  .withMessage("invalid application id. (Must be a Mongo ObjectID.)")
+  .trim();
+
+const validateUserId = body("userId")
+  .isString()
+  .withMessage("userId must be a string.")
+  .trim()
+  .notEmpty()
+  .withMessage("userId must be a non-empty string.");
+
+const validateCompanyId = body("companyId")
+  .isMongoId()
+  .withMessage("invalid company ID. (Must be a Mongo ObjectID.)")
+  .trim();
+
+const validatePosition = body("position")
+  .isString()
+  .withMessage("position must be a string.")
+  .trim()
+  .notEmpty()
+  .withMessage("position must be a non-empty string.");
+
+const validateCompanyName = body("companyName")
+  .isString()
+  .withMessage("company name must be a string.")
+  .trim()
+  .notEmpty()
+  .withMessage("company name must be a non-empty string.");
+
+const validateLink = body("link")
+  .optional()
+  .isURL({
+    require_valid_protocol: true,
+  })
+  .withMessage("link must be a valid URL.")
+  .trim();
+
+const validateProgress = [
+  body("progress")
+    .optional()
+    .isArray()
+    .withMessage("progress must be an array of application statuses."),
+  body("progress.*.status")
+    .isString()
+    .withMessage("Each application status must be a string.")
+    .trim()
+    .notEmpty()
+    .withMessage("each application status must be a non-empty string."),
+  body("progress.*.date")
+    .optional()
+    .notEmpty()
+    .withMessage("each application status must have a date.")
+    .bail()
+    .isISO8601({ strict: true })
+    .withMessage(
+      "each application status's date must be a valid ISO 8601 date.",
+    ),
+];
```

Comment:
Apologies for the sudden change of requirements, could you add a `note` field to `ApplicationProgress` as well. And also could you make the `status` field an enum of "APPLIED", "OA", "PHONE", "FINAL", "OFFER", "REJECTED". I think it might make more sense to have status as a stricter field but allow the user to add any additional notes to the progress.